### ISSUE TITLE
Rails 5 API support

### ISSUE THREAD

**Issue by ivanjolic95** [thoughtbot/ember-cli-rails#376]
Are there any plans to support Rails 5 API? The problem with it is that API controllers don't have helper_method. One solution would be to inherit gem's EmberController from a special EmberController (or whatever the name is) in app that inherits ActionController::Base instead of ActionController::API. Right now it inherits ApplicationController which is an API controller in Rails 5 API. I've configured ember-cli-rails to support testing the gem on Rails 5.0.0.beta1 locally so I can make it work that way.

**Comment by ivanjolic95** [comment]
The other solution would be to add a note in readme to make ApplicationController inherit ActionController::Base when using Rails 5 API.

**Comment by seanpdoyle** [comment]
@ivanjolic95 thanks for opening this issue.

I'm unfamiliar with most of the changes coming in Rails 5, so I'm having trouble following your suggestions.

Are you saying that when you `rails new --api`, your app's `ApplicationController` extends from `ActionController::API` instead of `ActionController::Base`?

**Comment by ivanjolic95** [comment]
Yeah, that's right.

**Comment by ivanjolic95** [comment]
And if you try to use this gem with it, there's an error because helper_method is not defined for the ActionController::API

**Comment by KirillSuhodolov** [comment]
@ivanjolic95 @seanpdoyle I updated my project with rails 5 beta 1. I am using full rails(not only API) and I get 500 standard error instead of ember application and nothing in server logs.
It looks like ActionController::Base really doesn't have helper_method

**Comment by ivanjolic95** [comment]
@KirillSuhodolov it's working for me when I use ActionController::Base. Which version of ember are you using?

**Comment by CollinGraves** [comment]
If anyone is following this issue and needs ember-cli-rails support with a Rails 5 build, feel free to use [my fork of this repository](http://github.com/CollinGraves/ember-cli-rails) until this gets straightened out. I'll keep it as current as possible. 

Thoughts on this going forward, @seanpdoyle? Would love to send a patch through to you.

**Comment by seanpdoyle** [comment]
https://github.com/thoughtbot/ember-cli-rails/pull/400 attempts to fix this issue.

**Comment by seanpdoyle** [comment]
@CollinGraves @ivanjolic95 @KirillSuhodolov https://github.com/thoughtbot/ember-cli-rails/pull/400 supports Rails `5.0.0.beta2`.

Would you mind seeing if it resolves your issues?:

```rb
# Gemfile

gem "ember-cli-rails", github: "thoughtbot/ember-cli-rails", branch: "sd-rails-5"
```

Unfortunately, Rails 5 depends on `rack@2.0.x`, which depends on `>= ruby@2.2.2`.

Dropping support for older Rubies constitutes a breaking change, and according to semantic versioning will require a major bump (in our case, a minor bump since the project is pre `1.0.x`).

This will likely delay the release of Rails 5 support. Before we introduce a breaking change, I'd like to wait until the stable release of Rails 5 (i.e. no `-beta`).

Does that sound reasonable?

**Comment by makabde** [comment]
@seanpdoyle I have tried your above solution, using the `sd-rails-5` branch with a newly generated Rails 5 API application, and a newly generated ember application. When booting the Rails app with either `bin/rails s` or `be rails s` the index page stays blank, but I have the following console output

```
=> Booting Puma
=> Rails 5.0.0.beta2 application starting in development on http://localhost:3000
=> Run `rails server -h` for more startup options
=> Ctrl-C to shutdown server
Puma 2.16.0 starting...
* Min threads: 0, max threads: 16
* Environment: development
* Listening on tcp://localhost:3000
Started GET "/" for ::1 at 2016-02-10 15:54:58 +0100
Processing by EmberCli::EmberController#index as HTML
  Parameters: {"ember_app"=>:client}
Completed 200 OK in 0ms (Views: 0.1ms | ActiveRecord: 0.0ms)
```

I have followed the setup instructions and did not customize any view since the base Rails app is an API only configured Rails app.

**Comment by seanpdoyle** [comment]
@makabde thanks for the additional information.

Just to clarify, in your application, `ApplicationController < ActionController::API`?

**Comment by makabde** [comment]
@seanpdoyle yes, that right.

**Comment by seanpdoyle** [comment]
@makabde thanks!

> the index page stays blank

Could you share the HTML that is returned? `Processing by EmberCli::EmberController#index as HTML` gives me reason to believe that you're getting some sort of HTML response.

Are there any `404` responses in your browser's network tab?

**Comment by makabde** [comment]
@seanpdoyle I created a repo so that you can maybe clone it and try yourself. So far both, the Rails and the Ember app are new but I would still expect to see the "Welcome to Ember", if everything was working as expected.

https://github.com/makabde/tapp

Also some screenshots of the returned HTML and of the browsers network tab.
<img width="1437" alt="screen shot 2016-02-10 at 16 39 57" src="https://cloud.githubusercontent.com/assets/801908/12952136/b7c83ef6-d015-11e5-984b-347d592121ec.png">
<img width="1437" alt="screen shot 2016-02-10 at 16 40 10" src="https://cloud.githubusercontent.com/assets/801908/12952137/b7ca2202-d015-11e5-94dd-2d6ce6afd64c.png">

**Comment by seanpdoyle** [comment]
The [`sd-rails-5`](https://github.com/thoughtbot/ember-cli-rails/tree/sd-rails-5) branch should now support your use case, with some side-effects that might be unfortunately unavoidable.

@makabde Please see [this new commit](https://github.com/thoughtbot/ember-cli-rails/commit/a9c799ca6f3f13f051221bcc1d17fef4c52a76bd).

**Comment by makabde** [comment]
@seanpdoyle The application now runs as expected, with this new commit. Do you think you will be able to come up with a solution that you will somehow prevent those side-effect you were referencing [in the commit](https://github.com/thoughtbot/ember-cli-rails/commit/a9c799ca6f3f13f051221bcc1d17fef4c52a76bd).

On the other hand would you recommend sticking to a "regular" Rails app with a dedicated controller for the API that will inherit from `ActionController::API`; in order to avoid those potentially annoying side effects?

**Comment by seanpdoyle** [comment]
@makabde [this commit](https://github.com/thoughtbot/ember-cli-rails/commit/1ccad80010c5a82812f3c6a1dc1517f3caf50aba) attempts to avoid `< ActionController::Base` by:

* inheriting from `ApplicationController`
* re-including modules into `EmberCli::EmberController` that are in `ActionController::Base`, but not in `ActionController::API`

This **feels** like an awful idea, but is effective in leveraging application-wide controller logic shared in `ApplicationController`.

**Comment by makabde** [comment]
:thumbsup:

**Comment by quezacoatl** [comment]
I get the following error:
```
ActionView::Template::Error (undefined method `render_ember_app' for #<#<Class:0x00000003a43438>:0x00000003ae0f08>):
    1: <%= render_ember_app params[:ember_app] do |head| %>
    2:   <% head.append do %>
    3:     <%= csrf_meta_tags %>
    4:   <% end %>
```

even when using `sd-rails-5` branch. I'm using `rails/master` too. It works if I change to `< ActionController::Base`, but then I don't see any point in the  `sd-rails-5` branch. Any other solutions?

**Comment by williamweckl** [comment]
Any updates?

**Comment by undergroundwebdesigns** [comment]
FYI, to those interested, I've just submitted #466 which fixes the "undefined method `render_ember_app'" error and gets the branch `sd-rails-5` working as expected on Rails 5 rc2.

**Comment by williamweckl** [comment]
Rails 5 launched and ember-cli-rails does not support it... :(

ember-cli-rails was resolved to 0.7.4, which depends on
      railties (< 5, >= 3.2)

**Comment by undergroundwebdesigns** [comment]
@williamweckl My fork https://github.com/undergroundwebdevelopment/ember-cli-rails/tree/sd-rails-5 is working for me, and includes the railties dependency bump.

**Comment by williamweckl** [comment]
@undergroundwebdesigns ty :)

**Comment by hauleth** [comment]
What is status of this? Rails 5 is now stable.

**Comment by seanpdoyle** [comment]
Hi @hauleth!

We've made some serious progress on Rails 5 support in [this PR](https://github.com/thoughtbot/ember-cli-rails/pull/400), but we're currently blocked by some JRuby test failures.

We could use some help, as I don't have any personal experience with JRuby.

Once those tests are passing, we'll be able to merge and release Rails 5 support.

**Comment by hauleth** [comment]
@seanpdoyle seems like this is related to RVM and JRuby in general, not to this gem jruby/jruby#2496.

**Comment by seanpdoyle** [comment]
closed by #400
